### PR TITLE
Generics

### PR DESCRIPTION
**Pull request by Manu0912**
# Summary
In this pull request i implemented generics to save in a container class a lis of Clients, Cars or Rentals. Also i created Client and Rental classes.

## Specifics
- 3 Classes have been created: Client.cs, Rental.cs and Container.cs
- Implement Generics in Container.cs to replace CarCRUD.cs
- Delete CarCRUD.cs
- New routes in appsettings.json
- Implement the function GetAll()
- implement the function GetFilePath()

### CODE REVIEW

File: AgenciaAutos/AgenciaAutos/Classes/Car.cs [hunk 1]
```diff
@@ -1,22 +1,21 @@
-﻿namespace AgenciaAutos.car 
+﻿using AgenciaAutos.Classes;
```
Comment (Manu0912) [inline AgenciaAutos/AgenciaAutos/Classes/Car.cs:1]: @jrocaspana

File: AgenciaAutos/AgenciaAutos/Classes/Car.cs [hunk 2]
```diff
@@ -1,22 +1,21 @@
-﻿namespace AgenciaAutos.car
```
Comment (Manu0912) [inline AgenciaAutos/AgenciaAutos/Classes/Car.cs:1]: @jrocaspana

File: AgenciaAutos/AgenciaAutos/Classes/Car.cs [hunk 3]
```diff
@@ -1,22 +1,21 @@
...
-    public class Car
+    public class Car : IEntity
     {
-        public int id { get; set; }
+        public int Id { get; set; }
         public string model { get; set; }
         public int doorsNumber { get; set; }
         public string color { get; set; }
         public string brand { get; set; }
         public string transmition { get; set; }
```
Comment (jrocaspana) [inline AgenciaAutos/AgenciaAutos/Classes/Car.cs:12]: this should be all PascalCase

File: AgenciaAutos/AgenciaAutos/Classes/Car.cs [hunk 4]
```diff
@@ -1,22 +1,21 @@
...
 
-        public Car()
-        {
-
-        }
+        public Car() { }
         public Car(string _model, int _doorsNumber, string _color, string _brand, string _transmition, int _id)
```
Comment (jrocaspana) [inline AgenciaAutos/AgenciaAutos/Classes/Car.cs:15]: this should be all camelCase (without `_` prefix)

File: AgenciaAutos/AgenciaAutos/Classes/Client.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        public string lastName { get; set; }
+        public string telephoneNumber { get; set; }
+        public string address { get; set; }
+        public string city { get; set; }
+        public string state { get; set; }
+        public int postalCode { get; set; }
+        public DateTime lastModified { get; set; }
+
+        public Client() { }
+        public Client(int _id,string _name, string _lastname, string _telephoneNumber, string _address, string _city, string _state, int _postalCode, DateTime _lastModified)
```
Comment (jrocaspana) [inline AgenciaAutos/AgenciaAutos/Classes/Client.cs:19]: same comment applies here, watch out for casing issues and always try to follow conventions!

File: AgenciaAutos/AgenciaAutos/Classes/Container.cs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+                    if (item.Id == x.Id)
+                    {
+                        newList.Add(item);
+                        obj = item;
+                    }
+                    else
+                    {
+                        newList.Add(x);
+                    }
+            });
```
Comment (jrocaspana) [inline AgenciaAutos/AgenciaAutos/Classes/Container.cs:86]: technically this is not _wrong_ but I think a normal foreach loop would read a lot better.
also, let's see if I understand the logic correctly: you are creating a new list based on the existing list + the new item, right? what if the entity isn't found?
also, did you try just updating the properties of the entity when found?

File: AgenciaAutos/AgenciaAutos/Classes/Container.cs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+            return list;
+        }
+
+        public string GetFilePath(string basePath)
+        {
+            StringBuilder sb = new();
+
+            sb.AppendFormat("{0}{1}{2}", basePath, typeof(T).Name,".json");
+
+            return sb.ToString();
```
Comment (jrocaspana) [inline AgenciaAutos/AgenciaAutos/Classes/Container.cs:144]: this is not wrong, but using string interpolation would be OK too (StringBuilder is intended to be use when dealing with really big strings)

File: AgenciaAutos/AgenciaAutos/Classes/Entity.cs [hunk 8]
```diff
@@ -0,0 +1,8 @@
+﻿
+namespace AgenciaAutos.Classes
+{
+    public interface IEntity
+    {
+        public int Id { get; set; }
+    }
```
Comment (jrocaspana) [inline AgenciaAutos/AgenciaAutos/Classes/Entity.cs:7]: niceeeeeeee

File: AgenciaAutos/AgenciaAutos/Classes/Rental.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    class Rental : IEntity
+    {
+        public int Id { get; set; }
+        public string rentalDuration { get; set; }
+        public string clientDni { get; set; }
+        public string carRentalId { get; set; }
+        public DateTime returnedDate { get; set; }
+
+        public Rental() { }
+        public Rental(int _id, string _rentalDuration, string _clientDni, string _carRentalId, DateTime _returnedDate)
```
Comment (jrocaspana) [inline AgenciaAutos/AgenciaAutos/Classes/Rental.cs:15]: same comment about casing and conventions apply here!!

**Review by jrocaspana — COMMENTED** [review]
sorry about the delay! this looks really good, let's just work on casing, conventions and try to come up with a better (performance wise) implementation of the update method

File: AgenciaAutos/AgenciaAutos/Classes/Client.cs (hunk 5 above)
Comment (Manu0912) [reply AgenciaAutos/AgenciaAutos/Classes/Client.cs:19]: yes, i´m trying to adapt me to write with c# conventions.

File: AgenciaAutos/AgenciaAutos/Classes/Container.cs (hunk 6 above)
Comment (Manu0912) [reply AgenciaAutos/AgenciaAutos/Classes/Container.cs:86]: I create a new list with the elements of the old list but if the current element in the foreach  is that i wanna change i put the updated element in its place.

File: AgenciaAutos/AgenciaAutos/Classes/Container.cs (hunk 7 above)
Comment (Manu0912) [reply AgenciaAutos/AgenciaAutos/Classes/Container.cs:144]: okey, i will change it. Thank you!

**Pull request merged**